src/server/journal/serializer.cc

```diff
@@ -145,13 +145,35 @@ io::Result<size_t> JournalReader::ReadString(io::MutableBytes buffer) {
   size_t size = 0;
   SET_OR_UNEXPECT(ReadUInt<uint64_t>(), size);
 
-  if (auto ec = EnsureRead(size); ec)
-    return make_unexpected(ec);
-
   if (size > buffer.size())
     return make_unexpected(make_error_code(errc::bad_message));
 
-  buf_.ReadAndConsume(size, buffer.data());
+  uint64_t available = std::min(size, buf_.InputLen());
+  uint64_t remainder = 0;
+
+  if (available < size) {
+    remainder = size - available;
+  }
+
+  buf_.ReadAndConsume(available, buffer.data());
+
```

Comment:
nit wrap lines 160-176  in `if (remainder)` condition - I think it will improve the readability of the code and simplify the conditions inside.